pubsub/alicloud/confluentkafka/confluentkafka.go

```diff
@@ -0,0 +1,266 @@
+	Ctx        context.Context
+	CancelFunc context.CancelFunc
+}
+
+type Kafka struct {
+	ConfigMap *kafka.ConfigMap
+	logger    logger.Logger
+	producer  *kafka.Producer
+	consumer  *kafka.Consumer
+	topics    []string
+
+	consumerContext contextWithCancelFunc
+}
+
+func NewConfluentKafka(logger logger.Logger) pubsub.PubSub {
+	return &Kafka{ //nolint:exhaustivestruct
+		logger: logger,
+	}
+}
+
+func (k *Kafka) parseMetadata(props map[string]string) error {
+	k.ConfigMap = &kafka.ConfigMap{}
+
+	var metadata map[string]interface{}
+	if configJson, ok := props["configJson"]; ok {
+		err := json.Unmarshal([]byte(configJson), &metadata)
+		if err != nil {
+			return err
+		}
+
+		for key, val := range metadata{
+			if reflect.ValueOf(val).Kind() == reflect.Float64{
+				err = k.ConfigMap.SetKey(key, int(val.(float64)))
+				if err != nil {
+					return err
+				}
+			} else {
+				err = k.ConfigMap.SetKey(key, val)
+				if err != nil {
+					return err
+				}
+			}
+		}
+
+		return nil
+	} else {
+		return fmt.Errorf("needs key configJson in metadata for initialization and check https://github.com/confluentinc/confluent-kafka-go for format of the config")
+	}
+}
+
+func (k *Kafka) Init(md pubsub.Metadata) error {
+	err := k.parseMetadata(md.Properties)
+	if err != nil {
+		return err
+	}
+
+	k.producer, err = kafka.NewProducer(k.ConfigMap)
+	if err != nil {
+		return err
+	}
+
+	k.consumer, err = kafka.NewConsumer(k.ConfigMap)
+	if err != nil {
+		return err
+	}
+
+	k.topics = make([]string, 0)
+
+	return nil
+}
+
+func parsePublishMetadata(md map[string]string, msg *kafka.Message) error {
+	if partitionStr, ok := md["partition"]; ok {
+		partition, err := strconv.ParseInt(partitionStr, 10, 32)
+		if err != nil {
+			return err
+		}
+		msg.TopicPartition.Partition = int32(partition)
+	} else {
+		msg.TopicPartition.Partition = kafka.PartitionAny
+	}
+
+	if offsetStr, ok := md["offset"]; ok {
+		offset, err := strconv.ParseInt(offsetStr, 10, 64)
+		if err != nil {
+			return err
+		}
+		msg.TopicPartition.Offset = kafka.Offset(offset)
+	}
+
+	if keyStr, ok := md["key"]; ok {
+		msg.Key = []byte(keyStr)
+	}
+
+	// if timestampStr, ok := md["timestamp"]; ok {
+	// 	if timeLayout, ok := md["timeLayout"]; ok {
+	// 		var err error
+	// 		msg.Timestamp, err = time.Parse(timeLayout, timestampStr)
+	// 		if err != nil {
+	// 			return err
+	// 		}
+	// 	} else {
+	// 		return fmt.Errorf("a value associated with key timeLayout must be provided for parsing timestamp, for more info: https://golang.org/src/time/format.go")
+	// 	}
+
+	// 	if timestampTypeStr, ok := md["timestampType"]; ok {
+	// 		timestampType, err := strconv.Atoi(timestampTypeStr)
+	// 		if err != nil {
+	// 			return err
+	// 		}
+	// 		msg.TimestampType = kafka.TimestampType(timestampType)
+	// 	} else {
+	// 		return fmt.Errorf("a value associated with key timestampType must be provided if timestamp is provided")
+	// 	}
+	// }
+
+	if headerJson, ok := md["headers"]; ok {
```

Comment:
IMO encoding the messages as JSON seems like an unnecessary step...  wondering if it should any metadata fields that are not handled above (key, partition) or ones that have a name with a prefix of `header:`.

Reply:
Changed to the prefix way